gcc 7.3.0 (new formula)

xu-cheng opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Linuxbrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Linuxbrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This adds gcc 7.x support to the Linuxbrew. 

Further, I think some changes made in this formula can be backported to the gcc-5 and gcc-6 formulae. 

Compared with the existing gcc formula, some of the major changes are:

 * Pass the specs to ./configure so that gcc can pickup brewed glibc.
   This fixes the building failure if the building system uses brewed gcc and brewed glibc
 * Use the `--libdir` flag to tell gcc to install libraries under versioned path.
   This is essential to allow different version of gcc formulae to be linked into
   the HOMEBREW_PREFIX.
   However, such flag should be only passed for the `gcc@*` versioned formulae.
   For the main formula `gcc`, we want its libraries to be linked to the
   unversioned path. So the bottles can be poured regardless the version of gcc.
 * Revamp the spec file. Specifically, use `-isysroot /nonexistent` and `-idirafter <dir>`
   to instruct gcc to search its internal header files before the system ones.

xu-cheng commented:
```
gcc@7:
  * Non-libraries were installed to "/home/linuxbrew/.linuxbrew/opt/gcc@7/lib"
    Installing non-libraries to "lib" is discouraged.
    The offending files are:
      /home/linuxbrew/.linuxbrew/opt/gcc@7/lib/libgfortran.spec
            /home/linuxbrew/.linuxbrew/opt/gcc@7/lib/libitm.spec
            /home/linuxbrew/.linuxbrew/opt/gcc@7/lib/libsanitizer.spec
            /home/linuxbrew/.linuxbrew/opt/gcc@7/lib/libcilkrts.spec
            /home/linuxbrew/.linuxbrew/opt/gcc@7/lib/libgomp.spec
            /home/linuxbrew/.linuxbrew/opt/gcc@7/lib/libmpx.spec
==> FAILED
```

This is the only error, which I think we will just ignore?

iMichka commented:
Hi. Really cool.

We can ignore the audit failure I think. If it causes trouble we can see what to do later.

I am travelling right now, I will review this as soon as possible.

I am ok to backport the changes to gcc6 and gcc5.

sjackman commented:
Excellent work, Xu! Thanks! Could you split this PR into two commits? One that copies the upstream `gcc` formula exactly, and a second commit to make your changes.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -0,0 +1,263 @@
+          # Pass the specs to ./configure so that gcc can pickup brewed glibc.
+          # This fixes the building failure if the building system uses brewed gcc
+          # and brewed glibc. There is very limit document on `--with-specs` option.
+          # But some relevant information can be found at https://stackoverflow.com/a/47911839
```
Link to this documentation instead:
https://gcc.gnu.org/onlinedocs/gcc/Spec-Files.html

xu-cheng replied:
I added this document as well. However, it is worth noting that `--with-specs` is a little different than regular spec files. That is why I put that stackoverflow url for reference.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -0,0 +1,263 @@
+  end
+
+  bottle do
+    cellar :any unless OS.mac? # gcc is designed to be portable.
```
Although technically correct `brew bottle` has trouble parsing this line, so I've stopped doing it. Use simply `cellar :any` instead. You can remove the bottle block entirely for this PR.

sjackman commented:
> The most important change is the passing of the `--with-specs` option to the configure. This is especially important when building the formula using brewed gcc and brewed glibc. Currently, building gcc-6 from source using brewed gcc compiler will failure due to unable to find the glibc during the link step.

Since they're `cellar :any`, they can be installed from the bottle, which is built on CircleCI using `/usr/bin/gcc`, so the `gcc@*` should not ever be built using a brewed `gcc`. It would simplify the formula to remove this entire stanza `if build.with? "glibc"`, because for the same reason, `gcc` shouldn't need to be built against the brewed `glibc`. It's better I think to design the formula so that the same compiler toolchain is built whether it uses the host `glibc` or the brewed `glibc`.

xu-cheng commented:
I have split the commits and made the requested changes. 

About the `--with-specs`, my main goal is to make the formula buildable from the source for either brewed gcc/glibc or system gcc/glibc if they are newly enough.  Therefore, it covers all user cases.
Accidentally, current CI tests cover exactly them, where CirceCI is using system gcc and TravisCI uses brewed one.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -74,34 +82,66 @@ def install
+        binutils = Dir["#{Formula["binutils"].opt_prefix}/*/bin"].first
+        glibc = Formula["glibc"].opt_lib
+        args += [
+          "--with-native-system-header-dir=#{HOMEBREW_PREFIX}/include",
```
I'm thinking this line should instead be
```ruby
"--with-native-system-header-dir=#{Formula["glibc"].include}",
```

sjackman replied:
This comment was ill advised. `gcc` needs to find the headers of both `linux-headers` and `glibc` in this directory.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -74,34 +82,66 @@ def install
+        args += [
+          "--with-native-system-header-dir=#{HOMEBREW_PREFIX}/include",
+          "--with-local-prefix=#{HOMEBREW_PREFIX}/local",
+          "--with-build-time-tools=#{binutils}",
```
I'd like to remove this line.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -74,34 +82,66 @@ def install
+        glibc = Formula["glibc"].opt_lib
+        args += [
+          "--with-native-system-header-dir=#{HOMEBREW_PREFIX}/include",
+          "--with-local-prefix=#{HOMEBREW_PREFIX}/local",
```
I'm thinking this line should be included in the configuration, with out without brewed `glibc`, to prevent including `/usr/local/include` in the default search path. See https://gcc.gnu.org/install/configure.html
@iMichka Sound good?

xu-cheng replied:
I copied this line from the existing `gcc` formula. But should it be `--with-local-prefix=#{HOMEBREW_PREFIX}`, as `#{HOMEBREW_PREFIX}/local` does not exist at all. Agreed that it should be included as default.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -74,34 +82,66 @@ def install
+          "--with-specs=%{!static:%x{--dynamic-linker=#{HOMEBREW_PREFIX}/lib/ld.so} %x{-rpath=#{HOMEBREW_PREFIX}/lib}}",
+        ]
+        # Set the search path for glibc libraries and objects.
+        ENV["LIBRARY_PATH"] = glibc
```
I'd like to remove this line, and see whether it works without it. I think that Superenv should take care of this.

xu-cheng replied:
Looks like that it is required.

https://travis-ci.org/Linuxbrew/homebrew-core/builds/355217535#L1910

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -148,7 +239,7 @@ def add_suffix(file, suffix)
     EOS
     system "#{bin}/gcc-#{version_suffix}", "-o", "hello-c", "hello-c.c"
-    assert_equal "Hello, world!\n", `./hello-c`
+    assert_equal "Hello, world!\n", shell_output("./hello-c")
```
Please send these changes upstream to Homebrew/core and remove them from this PR.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -159,7 +250,7 @@ def add_suffix(file, suffix)
     EOS
     system "#{bin}/g++-#{version_suffix}", "-o", "hello-cc", "hello-cc.cc"
-    assert_equal "Hello, world!\n", `./hello-cc`
+    assert_equal "Hello, world!\n", shell_output("./hello-cc")
```
Ditto.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -172,7 +263,7 @@ def add_suffix(file, suffix)
-    system "#{bin}/gfortran", "-o", "test", "test.f90"
-    assert_equal "Done\n", `./test`
+    system "#{bin}/gfortran-#{version_suffix}", "-o", "test", "test.f90"
+    assert_equal "Done\n", shell_output("./test")
```
Upstream

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -172,7 +263,7 @@ def add_suffix(file, suffix)
     EOS
-    system "#{bin}/gfortran", "-o", "test", "test.f90"
-    assert_equal "Done\n", `./test`
+    system "#{bin}/gfortran-#{version_suffix}", "-o", "test", "test.f90"
```
Is `-#{version_suffix}` necessary?

sjackman replied:
Bump.

xu-cheng replied:
Yes, it is. The non-versioned binaries are only presented in the main gcc formula.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -42,22 +48,24 @@ def version_suffix
-      sha256 "f7772a6ba73f44a6b378e4fe3548e0284f48ae2d02c701df1be93780c1607074"
+      url "https://raw.githubusercontent.com/Homebrew/formula-patches/e9e0ee09389a54cc4c8fe1c24ebca3cd765ed0ba/gcc/6.1.0-jit.patch"
+      sha256 "863957f90a934ee8f89707980473769cff47ca0663c3906992da6afb242fb220"
+    end
```
To make this patch smaller and easier to diff and read, can you please make the first…
```ruby
patch do
…
end if OS.mac?
```
and the second
```ruby
if OS.mac? && MacOS.version >= :high_sierra
```

sjackman commented:
I've made a few comments. I'm on the whole happy with this PR.

sjackman commented:
> About the --with-specs, my main goal is to make the formula buildable from the source for either brewed gcc/glibc or system gcc/glibc if they are newly enough. Therefore, it covers all user cases.
Accidentally, current CI tests cover exactly them, where CirceCI is using system gcc and TravisCI uses brewed one.

Thanks, Xu. I'm okay with keeping this `--with-specs`. It can be userful to developers to be able to build GCC formulae with the brewed `gcc`. It shouldn't typically be used by the users though, who should use the precompiled bottle.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -74,34 +83,63 @@ def install
+      ]
+    else
+      args += [
+        "--with-local-prefix=#{HOMEBREW_PREFIX}",
```
Should be…
```ruby
"--with-local-prefix=#{HOMEBREW_PREFIX}/local"
```

xu-cheng replied:
But `#{HOMEBREW_PREFIX}/local` does not exist at all?

sjackman replied:
Yep, that's the intention. We want to avoid picking up random headers from the user's host `/usr/local/include`. This way we pick up stray headers from a non-existant directory.

xu-cheng replied:
Thanks for the explanation. However, it still does not seem to be right for me. Therefore, I made some digging and experiments.

To avoid this comment being folded, I list my findings in the comment below.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -74,34 +83,63 @@ def install
+          # But some relevant information can be found at https://stackoverflow.com/a/47911839
+          "--with-specs=%{!static:%x{--dynamic-linker=#{HOMEBREW_PREFIX}/lib/ld.so} %x{-rpath=#{HOMEBREW_PREFIX}/lib}}",
+        ]
+        ENV["LIBRARY_PATH"] = glibc.opt_lib
```
Please remove this line. It should be handled by Superenv.

xu-cheng replied:
I tried, and it results to the building failure. See https://travis-ci.org/Linuxbrew/homebrew-core/builds/355217535#L1910

sjackman replied:
That particular build looks successful to me. It has a green check mark. It also however says `#13787 failed` even though it's green. The linked line shows no error message. Can you please copy and paste the error message to this issue, and link again to it? Please add to the formula a comment such as…
```ruby
# Fix the error: foo bar baz
ENV["LIBRARY_PATH"] = glibc.opt_lib
```

xu-cheng commented:
To better understand the `--with-local-prefix` option, I did more digging and these are what I found.

* According to the https://gcc.gnu.org/install/configure.html, it is used to specify the installation directory for local include files other than the default `/usr/local/include`. So it seems to me that we should set it to `HOMEBREW_PREFIX` instead of an non-exist directory.
* I tried to build the gcc-7 locally with `--with-local-prefix=#{HOMEBREW_PREFIX}/local` and `--with-native-system-header-dir=#{Formula["glibc"].include}`. This leads to a building failure with following error message:
```
/home/comp/chengxu/usr/opt/glibc/include/bits/errno.h:24:11: fatal error: linux/errno.h: No such file or directory
 # include <linux/errno.h>
           ^~~~~~~~~~~~~~~
```
* If I build the gcc-7 with `--with-local-prefix=#{HOMEBREW_PREFIX}`, the include search path looks good to me as shown in the below. Noted that my `HOMEBREW_PREFIX` is `/home/comp/chengxu/usr/`. And you can find that there is neither `/usr/local/include` or `/usr/include` in the search path. Further, `HOMEBREW_PREFIX/include` is in the search path with the intended order.
```
$ echo | ~/usr/opt/gcc@7/bin/gcc-7 -E -x c++ - -v
Reading specs from /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/specs
COLLECT_GCC=/home/comp/chengxu/usr/opt/gcc@7/bin/gcc-7
Target: x86_64-pc-linux-gnu
Configured with: ../configure --with-local-prefix=/home/comp/chengxu/usr --with-bugurl=https://github.com/Linuxbrew/homebrew-core/issues --disable-multilib --with-native-system-header-dir=/home/comp/chengxu/usr/opt/glibc/include --with-specs='%{!static:%x{--dynamic-linker=/home/comp/chengxu/usr/lib/ld.so} %x{-rpath=/home/comp/chengxu/usr/lib}}' --prefix=/home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1 --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-7 --with-gmp=/home/comp/chengxu/usr/opt/gmp --with-mpfr=/home/comp/chengxu/usr/opt/mpfr --with-mpc=/home/comp/chengxu/usr/opt/libmpc --with-isl=/home/comp/chengxu/usr/opt/isl --enable-checking=release --with-pkgversion='Homebrew GCC 7.3.0_1' --disable-nls
Thread model: posix
gcc version 7.3.0 (Homebrew GCC 7.3.0_1)
COLLECT_GCC_OPTIONS='-E' '-v' '-mtune=generic' '-march=x86-64'
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/libexec/gcc/x86_64-pc-linux-gnu/7.3.0/cc1plus -E -quiet -v -D_GNU_SOURCE - -mtune=generic -march=x86-64
ignoring nonexistent directory "/home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../x86_64-pc-linux-gnu/include"
#include "..." search starts here:
#include <...> search starts here:
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0/x86_64-pc-linux-gnu
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0/backward
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include
 /home/comp/chengxu/usr/include
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/include
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include-fixed
 /home/comp/chengxu/usr/opt/glibc/include
End of search list.
# 1 "<stdin>"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "/home/comp/chengxu/usr/include/stdc-predef.h" 1 3
# 1 "<command-line>" 2
# 1 "<stdin>"
COMPILER_PATH=/home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/libexec/gcc/x86_64-pc-linux-gnu/7.3.0/:/home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/libexec/gcc/x86_64-pc-linux-gnu/7.3.0/:/home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/libexec/gcc/x86_64-pc-linux-gnu/:/home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/:/home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/
LIBRARY_PATH=/home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/:/home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../lib64/:/lib/../lib64/:/usr/lib/../lib64/:/home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-E' '-v' '-mtune=generic' '-march=x86-64'
```
* Compared with the `gcc.rb` formula in my system, its search path is in below. You can find `/usr/local/include` and `/usr/include` are incorrectly included in the search path. Further, `HOMEBREW_PREFIX/include` is incorrectly put in a higher priority than `gcc`'s internal include paths (This should contributes to the `-isystem HOMEBREW_PREFIX/include` in the specs file generated during `post_install`).
```
$ echo | gcc -E -x c++ - -v
Reading specs from /home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/specs
COLLECT_GCC=gcc
Target: x86_64-unknown-linux-gnu
Configured with: ../configure --prefix=/home/linuxbrew/.linuxbrew/Cellar/gcc/5.5.0_3  --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-5 --with-gmp=/home/linuxbrew/.linuxbrew/opt/gmp --with-mpfr=/home/linuxbrew/.linuxbrew/opt/mpfr --with-mpc=/home/linuxbrew/.linuxbrew/opt/libmpc --with-isl=/home/linuxbrew/.linuxbrew/opt/isl@0.18  --enable-stage1-checking --enable-checking=release --enable-lto --with-build-config=bootstrap-debug --disable-werror --with-pkgversion='Homebrew gcc 5.5.0_3' --with-bugurl=https://github.com/Homebrew/homebrew/issues --with-boot-ldflags='-static-libstdc++ -static-libgcc ' --disable-nls --disable-multilib
Thread model: posix
gcc version 5.5.0 (Homebrew gcc 5.5.0_3)
COLLECT_GCC_OPTIONS='-E' '-v' '-mtune=generic' '-march=x86-64'
 /home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../libexec/gcc/x86_64-unknown-linux-gnu/5.5.0/cc1plus -E -quiet -v -imultiarch x86_64-linux-gnu -iprefix /home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/ -D_GNU_SOURCE - -isystem /home/comp/chengxu/usr/include -mtune=generic -march=x86-64
ignoring nonexistent directory "/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../x86_64-unknown-linux-gnu/include"
ignoring duplicate directory "/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/../../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../include/c++/5.5.0"
ignoring duplicate directory "/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/../../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../include/c++/5.5.0/x86_64-unknown-linux-gnu"
ignoring duplicate directory "/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/../../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../include/c++/5.5.0/backward"
ignoring duplicate directory "/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/../../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/include"
ignoring nonexistent directory "/usr/local/include/x86_64-linux-gnu"
ignoring duplicate directory "/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/../../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/include-fixed"
ignoring nonexistent directory "/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/../../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../x86_64-unknown-linux-gnu/include"
ignoring nonexistent directory "/usr/include/x86_64-linux-gnu"
#include "..." search starts here:
#include <...> search starts here:
 /home/comp/chengxu/usr/include
 /home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../include/c++/5.5.0
 /home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../include/c++/5.5.0/x86_64-unknown-linux-gnu
 /home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../include/c++/5.5.0/backward
 /home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/include
 /home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/include-fixed
 /usr/local/include
 /home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/../../include
 /usr/include
End of search list.
# 1 "<stdin>"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "/home/comp/chengxu/usr/include/stdc-predef.h" 1 3 4
# 1 "<command-line>" 2
# 1 "<stdin>"
COMPILER_PATH=/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../libexec/gcc/x86_64-unknown-linux-gnu/5.5.0/:/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../libexec/gcc/
LIBRARY_PATH=/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/:/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/:/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../lib64/:/lib/../lib64/:/usr/lib/../lib64/:/home/comp/chengxu/usr/Cellar/gcc/5.5.0_3/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-E' '-v' '-mtune=generic' '-march=x86-64'
```
* Finally, I am trying to build the gcc-7 with `--with-local-prefix=#{HOMEBREW_PREFIX}/local` and `--with-native-system-header-dir=#{HOMEBREW_PREFIX}/local` and see what is its search path. I will report back when it is done.

sjackman commented:
```
error: linux/errno.h: No such file or directory # include <linux/errno.h>
```

Drat. `gcc` needs the headers of both `linux-headers` and `glibc`. It'd be nice if `--with-native-system-header-dir` could accept a colon-separated list of directories, but I doubt that it does.

sjackman commented:
> According to the https://gcc.gnu.org/install/configure.html, it is used to specify the installation directory for local include files other than the default /usr/local/include. So it seems to me that we should set it to HOMEBREW_PREFIX instead of an non-exist directory.

It's fine that `$HOMEBREW_PREFIX/local/include` doesn't exist. The user could create it if they wanted to, and put a file in there. It should generally be empty though. `/usr/local/include` should also generally be empty.

sjackman commented:
I suggest…
```
--with-local-prefix=$HOMEBREW_PREFIX/local
--with-native-system-header-dir=$HOMEBREW_PREFIX/include
```

xu-cheng commented:
I summaries the the gcc include search path under various cases in the below table.

<table>
<tr>
    <th>Compiler</th>
    <th>`include` search path</th>
  </tr>
  <tr>
    <td> gcc built with
<pre>
--with-native-system-header-dir=#{Formula["glibc"].opt_include}
--with-local-prefix=#{HOMEBREW_PREFIX}
</pre>
  </td>
    <td>
<pre>
 HOMEBREW_PREFIX/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0
 HOMEBREW_PREFIX/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0/x86_64-pc-linux-gnu
 HOMEBREW_PREFIX/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0/backward
 HOMEBREW_PREFIX/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include
 HOMEBREW_PREFIX/include
 HOMEBREW_PREFIX/Cellar/gcc@7/7.3.0_1/include
 HOMEBREW_PREFIX/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include-fixed
 HOMEBREW_PREFIX/opt/glibc/include
</pre>
    </td>
  </tr>
  <tr>
    <td> gcc built with
<pre>
--with-native-system-header-dir=#{HOMEBREW_PREFIX}/include
--with-local-prefix=#{HOMEBREW_PREFIX}/local
</pre>
</td>
    <td>
<pre>
 HOMEBREW_PREFIX/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0
 HOMEBREW_PREFIX/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0/x86_64-pc-linux-gnu
 HOMEBREW_PREFIX/Cellar/gcc-test@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0/backward
 HOMEBREW_PREFIX/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include
 HOMEBREW_PREFIX/Cellar/gcc@7/7.3.0_1/include
 HOMEBREW_PREFIX/Cellar/gcc@7/7.3.0_1/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include-fixed
 HOMEBREW_PREFIX/include
</pre>
    </td>
  </tr>
    <tr>
    <td> system gcc on Linux</td>
    <td>
<pre>
 /usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7
 /usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/x86_64-redhat-linux
 /usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../include/c++/4.4.7/backward
 /usr/local/include
 /usr/lib/gcc/x86_64-redhat-linux/4.4.7/include
 /usr/include
</pre>
    </td>
  </tr>
    <tr>
    <td> gcc-7 on macOS</td>
    <td>
<pre>
 /usr/local/Cellar/gcc/7.3.0_1/lib/gcc/7/gcc/x86_64-apple-darwin17.3.0/7.3.0/../../../../../../include/c++/7.3.0
 /usr/local/Cellar/gcc/7.3.0_1/lib/gcc/7/gcc/x86_64-apple-darwin17.3.0/7.3.0/../../../../../../include/c++/7.3.0/x86_64-apple-darwin17.3.0
 /usr/local/Cellar/gcc/7.3.0_1/lib/gcc/7/gcc/x86_64-apple-darwin17.3.0/7.3.0/../../../../../../include/c++/7.3.0/backward
 /usr/local/Cellar/gcc/7.3.0_1/lib/gcc/7/gcc/x86_64-apple-darwin17.3.0/7.3.0/include
 /usr/local/include
 /usr/local/Cellar/gcc/7.3.0_1/include
 /usr/local/Cellar/gcc/7.3.0_1/lib/gcc/7/gcc/x86_64-apple-darwin17.3.0/7.3.0/include-fixed
 /usr/include
 /System/Library/Frameworks
 /Library/Frameworks
</pre>
    </td>
  </tr>
</table>

I still think that building with `--with-native-system-header-dir=#{Formula["glibc"].opt_include} --with-local-prefix=#{HOMEBREW_PREFIX}` makes more sense. The reason is that the search path under this case is consistent with  gcc search path outside the Linuxbrew. Please noted that `HOMEBREW_PREFIX/include` should have a higher search priority than `GCC_CELLAR/include`. This is what happens in the system gcc and gcc building with  `--with-local-prefix=#{HOMEBREW_PREFIX}`. In the contrary, gcc building with `-with-local-prefix=#{HOMEBREW_PREFIX}/local` puts `HOMEBREW_PREFIX/include` in the lowest priority.

sjackman commented:
I am happy with `HOMEBREW_PREFIX/include` being in the final position. The GCC cellar should be first and take priority over `HOMEBREW_PREFIX/include`.

xu-cheng commented:
I have updated PR as you requested.  My previous comments are mainly because I think passing an nonexist path is weird.  As for this PR, my main goal is to enable me to use gcc 7 so I can build my project with C++17.

xu-cheng commented:
@sjackman 

I double checked the bottle built by the CircleCI, I am afraid that we may have to use `--with-local-prefix=#{HOMEBREW_PREFIX}` after all.

The gcc built by the CircleCI (which is what the majority users get) has the following include search path:
```
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0/x86_64-pc-linux-gnu
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0/backward
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/include
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/include-fixed
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/../../include
 /usr/include
```

You will find that `HOMEBEW_PREFIX/include` is missing. Noted that we could add it back using Spec file with `-isystem` (like what current gcc formula does) or `-I`, but in such way, the `HOMEBREW_PREFIX/include` will have higher search priority than gcc internal paths as shown in the below:
```
 /home/comp/chengxu/usr/include
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0/x86_64-pc-linux-gnu
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../include/c++/7.3.0/backward
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/include
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/include-fixed
 /home/comp/chengxu/usr/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/../../include
 /usr/include
```

xu-cheng commented:
Turns out that using `--with-local-prefix=#{HOMEBREW_PREFIX}` does not solve the problem either. The local prefix is hardcoded into the binary file, as result it cannot be changed and hence it is unportable.

xu-cheng commented:
I think I have solved the problem using following spec file:
```
*cpp_unique_options:
 + -idirafter #{HOMEBREW_PREFIX}/include -isysroot #{HOMEBREW_PREFIX}/local
```
* `-idirafter #{HOMEBREW_PREFIX}/include` will put ` #{HOMEBREW_PREFIX}/include` in the search path but with lower priority than gcc internal path.
* `-isysroot #{HOMEBREW_PREFIX}/local` will cause `/usr/include` to become `#{HOMEBREW_PREFIX}/local/usr/include`, thus prevents gcc searching `/usr/include`.

I also changed back to use `--with-local-prefix=#{HOMEBREW_PREFIX}/local` as discussed before.

After this PR is merged, I will submit PR to  backport relevant changes to gcc 5 and gcc 6 formulae.

sjackman commented:
Good sleuthing. I'd suggest
```
*cpp_unique_options:
 + -idirafter #{HOMEBREW_PREFIX}/include -isysroot #{HOMEBREW_PREFIX}
```
Could you please post the library search path with this change?

xu-cheng commented:
Updated. The search path is:

```
Reading specs from @HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/specs
COLLECT_GCC=@HOMEBREW_PREFIX@/opt/gcc@7/bin/gcc-7
Target: x86_64-pc-linux-gnu
Configured with: ../configure --with-local-prefix=/home/linuxbrew/.linuxbrew/local --with-bugurl=https://github.com/Linuxbrew/homebrew-core/issues --disable-multilib --prefix=/home/linuxbrew/.linuxbrew/Cellar/gcc@7/7.3.0_1 --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-7 --with-gmp=/home/linuxbrew/.linuxbrew/opt/gmp --with-mpfr=/home/linuxbrew/.linuxbrew/opt/mpfr --with-mpc=/home/linuxbrew/.linuxbrew/opt/libmpc --with-isl=/home/linuxbrew/.linuxbrew/opt/isl --enable-checking=release --with-pkgversion='Homebrew GCC 7.3.0_1' --disable-nls
Thread model: posix
gcc version 7.3.0 (Homebrew GCC 7.3.0_1)
COLLECT_GCC_OPTIONS='-E' '-v' '-mtune=generic' '-march=x86-64'
 @HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../libexec/gcc/x86_64-pc-linux-gnu/7.3.0/cc1 -E -quiet -v -imultiarch x86_64-linux-gnu -iprefix @HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/ - -idirafter @HOMEBREW_PREFIX@/include -isysroot @HOMEBREW_PREFIX@ -mtune=generic -march=x86-64
ignoring nonexistent directory "@HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../x86_64-pc-linux-gnu/include"
ignoring duplicate directory "@HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/../../lib/gcc/x86_64-pc-linux-gnu/7.3.0/include"
ignoring nonexistent directory "@HOMEBREW_PREFIX@/home/linuxbrew/.linuxbrew/local/include/x86_64-linux-gnu"
ignoring nonexistent directory "@HOMEBREW_PREFIX@/home/linuxbrew/.linuxbrew/local/include"
ignoring duplicate directory "@HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/../../lib/gcc/x86_64-pc-linux-gnu/7.3.0/include-fixed"
ignoring nonexistent directory "@HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/../../lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../x86_64-pc-linux-gnu/include"
ignoring nonexistent directory "@HOMEBREW_PREFIX@/usr/include/x86_64-linux-gnu"
ignoring nonexistent directory "@HOMEBREW_PREFIX@/usr/include"
#include "..." search starts here:
#include <...> search starts here:
 @HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/include
 @HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/include-fixed
 @HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/../../include
 @HOMEBREW_PREFIX@/include
End of search list.
# 1 "<stdin>"
# 1 "<built-in>"
# 1 "<command-line>"
# 31 "<command-line>"
# 1 "@HOMEBREW_PREFIX@/include/stdc-predef.h" 1 3 4
# 32 "<command-line>" 2
# 1 "<stdin>"
COMPILER_PATH=@HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../libexec/gcc/x86_64-pc-linux-gnu/7.3.0/:@HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../libexec/gcc/
LIBRARY_PATH=@HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/:@HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/:@HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../../lib64/:/lib/../lib64/:/usr/lib/../lib64/:@HOMEBREW_PREFIX@/Cellar/gcc@7/7.3.0_1/bin/../lib/gcc/x86_64-pc-linux-gnu/7.3.0/../../../:/lib/:/usr/lib/
COLLECT_GCC_OPTIONS='-E' '-v' '-mtune=generic' '-march=x86-64'

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -0,0 +1,272 @@
+class GccAT7 < Formula
+  desc "GNU compiler collection"
+  homepage "https://gcc.gnu.org/"
+  revision 1
```
Please delete this line, since `gcc@7` is a new formula to Linuxbrew.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -0,0 +1,272 @@
+  end
+
+  bottle do
+    cellar :any # gcc is designed to be portable.
```
Please move this comment above the `bottle do` line. The bottle-block parsing code is not terribly smart.

sjackman commented:
Two comments, after which this formula looks good to go.

xu-cheng commented:
Updated, thanks for the comments. I will submit PR to backport changes to the gcc 5/6 formulae.

xu-cheng commented:
One last change, I found I need to allow gcc to search `/usr/include` if user uses system glibc. Otherwise, gcc will fail to find files like `stdio.h`. I have updated accordingly. 

Further, I add some comments on the gcc search path and why we set these values in spec file.

xu-cheng commented:
> I found I need to allow gcc to search /usr/include if user uses system glibc. 

Then, I found out that in Ubuntu some glibc header files are not in /usr/include but in /usr/include/x86_64-linux-gnu. And this path seems not to be fixed among different distros.

Any idea how to solve this edge case? One thing I think is that it may be better and more simply if we enforce to use brew glibc at the time.

sjackman commented:
> One thing I think is that it may be better and more simply if we enforce to use brew glibc all the time.

I prefer to use the host `glibc` when possible. In particular, it's necessary if the host version of `glibc` is newer than the brewed version.

xu-cheng commented:
>  One reason it may not work and that I'd like to test, is if both gcc (version 5) and gcc@7 are installed, the gcc@7 may pick up the C++ headers of gcc (version 5) if it finds those first, before finding its own gcc@7 C++ headers.

Yes, that is exactly what will happen if we use `-isystem HOMEBREW_PREFIX/include`.

I end up to solve this problem by extracting native system header dirs using `echo | gcc -E -x c - -v`.

sjackman commented:
Rather than using `system_header_dirs`, how about…
```ruby
*cpp_unique_options:
+ -idirafter #{HOMEBREW_PREFIX}/include" -isysroot #{"-isysroot #{HOMEBREW_PREFIX}" if glibc.any_version_installed?}
```

sjackman commented:
```
gcc@7:
  * C: 25: col 3: dependency "isl" (line 25) should be put before dependency "mpfr" (line 24)
  * W: 230: col 16: `File.exists?` is deprecated in favor of `File.exist?`.
  * Non-libraries were installed to "/home/linuxbrew/.linuxbrew/opt/gcc@7/lib"
    Installing non-libraries to "lib" is discouraged.
    The offending files are:
      /home/linuxbrew/.linuxbrew/opt/gcc@7/lib/libgomp.spec
            /home/linuxbrew/.linuxbrew/opt/gcc@7/lib/libsanitizer.spec
            /home/linuxbrew/.linuxbrew/opt/gcc@7/lib/libgfortran.spec
            /home/linuxbrew/.linuxbrew/opt/gcc@7/lib/libitm.spec
            /home/linuxbrew/.linuxbrew/opt/gcc@7/lib/libmpx.spec
            /home/linuxbrew/.linuxbrew/opt/gcc@7/lib/libcilkrts.spec
```
Let's fix the first two, and ignore the last one.

xu-cheng commented:
> Rather than using system_header_dirs, how about…

This solution puts system header files (/usr/include) before the brewed ones (HOMEBREW_PREFIX/include), which is not what we want.

As for the audit warnings, I have fixed the first two as requested.

sjackman commented:
Here's the list of directories that need to be search
```
/home/linuxbrew/.linuxbrew/Cellar/gcc/5.5.0_1/include/c++/5.5.0
/home/linuxbrew/.linuxbrew/Cellar/gcc/5.5.0_1/include/c++/5.5.0/x86_64-unknown-linux-gnu
/home/linuxbrew/.linuxbrew/Cellar/gcc/5.5.0_1/include/c++/5.5.0/backward
/home/linuxbrew/.linuxbrew/Cellar/gcc/5.5.0_1/lib/gcc/x86_64-unknown-linux-gnu/5.5.0/include
/home/linuxbrew/.linuxbrew/Cellar/gcc/5.5.0_1/lib/gcc/x86_64-unknown-linux-gnu/5.5.0/include-fixed
/home/linuxbrew/.linuxbrew/opt/glibc/include
/home/linuxbrew/.linuxbrew/opt/linux-headers/include
/home/linuxbrew/.linuxbrew/include
/usr/include unless glibc.any_version_installed?
```

Or put another way
```ruby
libgcc = "#{lib}/gcc/x86_64-unknown-linux-gnu/5.5.0"
includecxx = "#{include}/c++/5.5.0"
<<~EOS
#{includecxx}
#{includecxx}/x86_64-unknown-linux-gnu
#{includecxx}/backward
#{libgcc}/include
#{libgcc}/include-fixed
#{HOMEBREW_PREFIX}/opt/glibc/include
#{HOMEBREW_PREFIX}/opt/linux-headers/include
#{HOMEBREW_PREFIX}/include
#{"/usr/include" unless glibc.any_version_installed?}
EOS
```

How about…
```ruby
<<~EOS
-isystem #{HOMEBREW_PREFIX}/opt/glibc/include
-isystem #{HOMEBREW_PREFIX}/opt/linux-headers/include
-idirafter #{HOMEBREW_PREFIX}/include
-isysroot /nonexistant
#{"-idirafter /usr/include" unless glibc.any_version_installed?}
EOS
```

sjackman commented:
Yes, I believe this may work.
```diff
diff --git a/a b/b
index f0c74d7..0d01388 100644
--- a/a
+++ b/b
@@ -5,27 +5,27 @@ Configured with: ../configure --prefix=/home/linuxbrew/.linuxbrew/Cellar/gcc/5.5
 Thread model: posix
 gcc version 5.5.0 (Homebrew gcc 5.5.0_2) 
 COLLECT_GCC_OPTIONS='-E' '-v' '-mtune=generic' '-march=x86-64'
- /home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../libexec/gcc/x86_64-unknown-linux-gnu/5.5.0/cc1plus -E -quiet -v -imultiarch x86_64-linux-gnu -iprefix /home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/ -D_GNU_SOURCE - -isystem /home/sjackman/.linuxbrew/include -mtune=generic -march=x86-64
+ /home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../libexec/gcc/x86_64-unknown-linux-gnu/5.5.0/cc1plus -E -quiet -v -imultiarch x86_64-linux-gnu -iprefix /home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/ -D_GNU_SOURCE - -isysroot /nonexistent -idirafter /home/sjackman/.linuxbrew/include -mtune=generic -march=x86-64
 ignoring nonexistent directory "/home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../x86_64-unknown-linux-gnu/include"
 ignoring duplicate directory "/home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/../../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../include/c++/5.5.0"
 ignoring duplicate directory "/home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/../../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../include/c++/5.5.0/x86_64-unknown-linux-gnu"
 ignoring duplicate directory "/home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/../../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../include/c++/5.5.0/backward"
 ignoring duplicate directory "/home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/../../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/include"
-ignoring nonexistent directory "/usr/local/include/x86_64-linux-gnu"
+ignoring nonexistent directory "/nonexistent/usr/local/include/x86_64-linux-gnu"
+ignoring nonexistent directory "/nonexistent/usr/local/include"
 ignoring duplicate directory "/home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/../../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/include-fixed"
 ignoring nonexistent directory "/home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/../../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../x86_64-unknown-linux-gnu/include"
-ignoring nonexistent directory "/usr/include/x86_64-linux-gnu"
+ignoring nonexistent directory "/nonexistent/usr/include/x86_64-linux-gnu"
+ignoring nonexistent directory "/nonexistent/usr/include"
 #include "..." search starts here:
 #include <...> search starts here:
- /home/sjackman/.linuxbrew/include
  /home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../include/c++/5.5.0
  /home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../include/c++/5.5.0/x86_64-unknown-linux-gnu
  /home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/../../../../include/c++/5.5.0/backward
  /home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/include
  /home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/x86_64-unknown-linux-gnu/5.5.0/include-fixed
- /usr/local/include
  /home/sjackman/.linuxbrew/Cellar/gcc/5.5.0_2/bin/../lib/gcc/../../include
- /usr/include
+ /home/sjackman/.linuxbrew/include
 End of search list.
 # 1 "<stdin>"
 # 1 "<built-in>"
```

sjackman commented:
With this scheme, there should be no need to set neither `--with-local-prefix` nor `--with-native-system-header-dir` since `-isysroot /nonexistent` makes both irrelevant.

xu-cheng commented:
> How about…

That is the exactly what I did in the first place. However, it turns out that `/usr/include` is not the only place to store the system header files. (As a result, `brew test gcc` on ubuntu fails). That's why I adopt the `system_header_dirs` logic.

 After some searching, I found out that the system header files can be located at these location based on the distro :
```
/usr/target/include
/usr/include/target
/usr/include
```
Therefore, I have simplified the logic accordingly.


> With this scheme, there should be no need to set neither --with-local-prefix nor --with-native-system-header-dir since -isysroot /nonexistent makes both irrelevant.

You are right. I have removed `--with-local-prefix`. `--with-native-system-header-dir` is still required as it is used in build time.

xu-cheng commented on Formula/gcc@7.rb:
```diff
@@ -0,0 +1,296 @@
+      osmajor = `uname -r`.chomp
+      args += [
+        "--build=x86_64-apple-darwin#{osmajor}",
+        "--libdir=#{lib}/gcc/#{version_suffix}",
```
@sjackman This argument is passed in the Homebrew to allow gcc and gcc@* both being linked to prefix instead of applying keg_only to the version formulae. Maybe we should do the same in Linuxbrew? Of cause, we will do something like this for main formula `gcc.rb`:
```
ln_sf Dir["#{lib}/gcc/#{version_suffix}/*"], lib
```

sjackman replied:
On Ubuntu I see libraries in `/usr/lib/x86_64-linux-gnu` and `/usr/lib` but nothing in `/usr/x86_64-linux-gnu/lib`. Do you see files in the latter directory?

On Redhat based distros it's `/usr/lib64` and `/usr/lib`. There may be magic sauce involved that causes `-L/usr/lib` to search both `/usr/lib64` and `/usr/lib`.

sjackman replied:
Yes, I would like to link `gcc` and `gcc@7` at the same time and avoid using `keg_only :versioned_formula`. We can adopt the Homebrew solution that you've suggested above. Another possible simple fix is to remove the block of code that moves GCC's files from `lib64` to `lib`. Since `lib64` is not symlinked into prefix by brew, there will be no conflict, and GCC already expects to find its files in `lib64`. Thoughts?

sjackman replied:
Using `lib64` may still require modifying the `specs` file to add `-rpath #{libdir}`. If that's the case, I'd probably rather go with the Homebrew solution. All in all, it's good to maintain consistency with Homebrew.

xu-cheng replied:
> On Ubuntu I see libraries in /usr/lib/x86_64-linux-gnu and /usr/lib but nothing in /usr/x86_64-linux-gnu/lib. Do you see files in the latter directory?
On Redhat based distros it's /usr/lib64 and /usr/lib. There may be magic sauce involved that causes -L/usr/lib to search both /usr/lib64 and /usr/lib.

Why do you ask these? We are not dealing with system lib here?  Or you are actually asking on the include header file?

xu-cheng replied:
> Using lib64 may still require modifying the specs file to add -rpath #{libdir}. If that's the case, I'd probably rather go with the Homebrew solution. All in all, it's good to maintain consistency with Homebrew.

Yes, I prefer this as well.

sjackman replied:
> Why do you ask these? We are not dealing with system lib here? Or you are actually asking on the include header file?

I meant the header files. I see no files in `/usr/x86_64-linux-gnu/include`, only in `/usr/include/x86_64-linux-gnu` and `/usr/include`.

sjackman replied:
> Yes, I prefer this as well.

Sounds good. Let's go ahead with that.

xu-cheng replied:
>Sounds good. Let's go ahead with #{lib}/gcc/#{version_suffix}

I just tried it, and it does not work as intended. By passing `—libdir=#{lib}/gcc/#{version_suffix}`, the shared libraries end up to be installed at `lib/gcc/version_suffix/lib64`.

So I checked the document, and find that there is a `—with-slibdir` argument specifically to set the location of these shared libaraies. As a result, the flags one become as the following:
```
—libdir=#{lib}   # so we don’t need the move lib64 to lib step 
—with-slibdir=#{lib}/gcc/#{version_suffix} # so the shared libaraies are now installed into the versioned path
```

sjackman replied:
> /usr/target/include is a path listed in some old gcc documents

I'd prefer not to include it unless we see it used in the wild.

sjackman replied:
> `—libdir=#{lib}` so we don’t need the move lib64 to lib step 

Does that work? Excellent if so! It may actually install libraries in `#{lib}/../lib64`, which I find frustrating. There will be a solution in the Debian patches.

> `—with-slibdir=#{lib}/gcc/#{version_suffix}` so the shared libaraies are now installed into the versioned path
 
This may not be necessary if we can solve the `--libdir` issue above.

xu-cheng replied:
> That's a cross-compiling package, so not relevant to native compilation.

I have removed /usr/targert/include path.

> -enable-version-specific-runtime-libs looks promising!

And it does not work as well. I finally made it work by using `inreplace "gcc/config/i386/t-linux64", "m64=../lib64", "m64="` plus the homebrew solution `"--libdir=#{lib}/gcc/#{version_suffix}"`.

xu-cheng commented:
After many trial and errors, I think I have made it work as intended. 

I think this PR is ready to be merged. Please check.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -0,0 +1,300 @@
+      libdir = HOMEBREW_PREFIX/"lib/gcc/#{version_suffix}"
+      specs.write specs_string + <<~EOS
+        *cpp_unique_options:
+        + -isysroot /nonexistent #{system_header_dirs.map { |p| "-idirafter #{p}" }.join(" ") }
```
```
==> brew audit gcc@7 --online --new-formula
==> Installing or updating 'rubocop' gem
Error: 1 problem in 1 formula
gcc@7:
  * C: 251: col 36: Space inside string interpolation detected.
==> FAILED
```
I think it may be complaining about the last space before the closing brace.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -74,11 +81,54 @@ def install
+      #
+      # However, such flag should be only passed for the `gcc@*` versioned formulae.
+      # For the main formula `gcc`, we want its libraries to be linked to the
+      # unversioned path. So the bottles can be poured regardless the version of gcc.
```
We can remove this comment. Thanks for the effort though. Please copy and paste this text into the GitHub issue.

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -138,6 +192,74 @@ def add_suffix(file, suffix)
+      specs_orig = Pathname.new("#{specs}.orig")
+      rm_f [specs_orig, specs]
+
+      system_header_dirs = %W[#{HOMEBREW_PREFIX}/include]
```
```ruby
system_header_dirs = ["#{HOMEBREW_PREFIX}/include]"]
```

sjackman commented on Formula/gcc@7.rb:
```diff
@@ -138,6 +192,74 @@ def add_suffix(file, suffix)
+        system_header_dirs += %W[
+          /usr/include/#{target}
+          /usr/include
+        ].select { |p| File.exist?(p) }
```
Please make this unconditional.
```ruby
system_header_dirs += ["/usr/include/#{target}", "/usr/include"]
```

sjackman commented:
Looking good! A few comments. We're getting very close.

xu-cheng commented:
Updated as requested.

sjackman commented:
Looks good to me! Thanks, Xu! Once CI finishes, I think this PR is good to merge. I'll test the resulting bottle myself before merging, to double-check that all is good.

Some time after Ubuntu 18.04 LTS is released, we'll bump the default versions of `glibc` and `gcc`. Would you like to submit a PR to bring these fixes to the `gcc@5` formula?

xu-cheng commented:
See #6955, #6956, #6957

xu-cheng commented:
> Some time after Ubuntu 18.04 LTS is released, we'll bump the default versions of glibc and gcc. 

A side note, this will also mean that the Linuxbrew will require kernel 3.2 comparing with current 2.6.16.
This will be a bad news for me as the clusters on my university uses a quite old version of the kernel.
```
$ uname -a
Linux csr50 2.6.32-696.20.1.el6.x86_64 #1 SMP Fri Jan 26 17:51:45 UTC 2018 x86_64 GNU/Linux
```

sjackman commented:
> A side note, this will also mean that the Linuxbrew will require kernel 3.2 comparing with current 2.6.16.

Thanks for the heads up. I hadn't yet looked up the minimum kernel version required for `glibc` that comes with Ubuntu 18.04. We'll discuss the issue and get back to you. The ability to run on HPC systems is definitely a design consideration of Linuxbrew.

iMichka commented:
@sjackman ready to merge on my side. You said you wanted to test the bottle before merging?

@xu-cheng Thank you for the tremendous work here and all the testing. This will be really helpfull!

sjackman commented:
I carried out the following test.
```sh
brew pull https://github.com/Linuxbrew/homebrew-core/pull/6649
brew install https://14950-56942570-gh.circle-artifacts.com/0/bottles/gcc%407-7.3.0.x86_64_linux.bottle.tar.gz
brew test gcc@7

brew install hello --cc=gcc-7
brew test hello
brew remove hello

brew install gcc
brew test gcc
brew test gcc@7

brew install hello --cc=gcc-7
brew test hello
brew remove hello

brew install hello --cc=gcc-5
brew test hello
brew remove hello
```
All is good!

Thanks for this fantastic work, Cheng! 💯
I'm good to merge.

iMichka commented:
Merged! Thanks again @xu-cheng !
